### PR TITLE
ghc 8.0 TH changes and warning flag changes

### PR DESCRIPTION
**Pull request by cartazio**
I believe these are all the changes needed for lens master to be clean building on ghc 8.0 tip as of today onwards

@ekmett  @hvr  @glguy  @bgamari

### CODE REVIEW

**Comment by cartazio** [comment]
i'm not sure if the CPP macros expose enough info to write this in a way that could support the release candidates that have come out so far at the same time as the tip of ghc-8.0 :(

**Comment by cartazio** [comment]
nb: the build failure seems to indicate that the current hvr ppa for 8.0.1 isn't the current tip of that branch

**Comment by hvr** [comment]
@cartazio yeah, @thoughtpolice suggested to keep the 8.0.1 package at RC3 state until RC4 is cut, that's why I'm holding back updating the 8.0.1 package.

**Comment by RyanGlScott** [comment]
@cartazio, can you add `Prism`s for `Overlap` in `Language.Haskell.TH.Lens` while you're at it?

**Comment by cartazio** [comment]
Good idea

**Comment by phadej** [comment]
I made a revisions for `lens-4.13.1` and `lens-4.13.2` to restrict `template-haskell-2.11`, I didn't touch `lens-4.13.2.1` as someone is probably is using older `ghc-8.0.1` RC and the package still works for them.

Please remember to make a revision after this is merged & released.

http://matrix.hackage.haskell.org/package/lens

**Comment by hvr** [comment]
fwiw, I restarted https://travis-ci.org/ekmett/lens/jobs/123983821 and now it passes

**Comment by phadej** [comment]
Now this is broken for `!= 8.0.*` :)

File: src/Language/Haskell/TH/Lens.hs [hunk 1]
```diff
@@ -848,12 +848,21 @@ _ClassD
       remitter (ClassD x y z w u) = Just (x, y, z, w, u)
       remitter _ = Nothing
 
-_InstanceD :: Prism' Dec (Cxt, Type, [Dec])
+#if MIN_VERSION_template_haskell(2,11,0)
+_InstanceD :: Prism' Dec (Maybe Overlap,Cxt, Type, [Dec])
+#else
+_InstanceD :: Prism' Dec (Maybe Overlap,Cxt, Type, [Dec])
```
Comment (hvr) [inline src/Language/Haskell/TH/Lens.hs]:
```
src/Language/Haskell/TH/Lens.hs:854:33:
    Not in scope: type constructor or class ‘Overlap’
```

**Comment by cartazio** [comment]
I'll do the cleanup this morning.   Thanks for being patient :)

File: src/Language/Haskell/TH/Lens.hs (hunk 1 above)
Comment (cartazio) [inline src/Language/Haskell/TH/Lens.hs]: Derp, good catch

**Comment by cartazio** [comment]
@RyanGlScott  what do you want the Prism to be between?  Just exposing the Sum Structure of the `Overlaps` data type?
```haskell
-- | Varieties of allowed instance overlap.
data Overlap = Overlappable   -- ^ May be overlapped by more specific instances
             | Overlapping    -- ^ May overlap a more general instance
             | Overlaps       -- ^ Both 'Overlapping' and 'Overlappable'
             | Incoherent     -- ^ Both 'Overlappable' and 'Overlappable', and
                              -- pick an arbitrary one if multiple choices are
                              -- available.
  deriving( Show, Eq, Ord, Data, Typeable, Generic )
```
or what?

File: src/Language/Haskell/TH/Lens.hs (hunk 1 above)
Comment (cartazio) [inline src/Language/Haskell/TH/Lens.hs]: should be fixed now i hope :)

**Comment by RyanGlScott** [comment]
The convention in `Language.Haskell.TH.Lens` is to create prisms of type `Prism Foo ()` for datatypes `Foo` which have all nullary constructors. [Here's an example](https://github.com/ekmett/lens/blob/b28e8cccdeec8f24bb83b153bfd0e1e227758d8b/src/Language/Haskell/TH/Lens.hs#L1272-L1294) for `data Safety = Safe | Unsafe | Interruptible`.

**Comment by cartazio** [comment]
@RyanGlScott should work now :)

**Comment by RyanGlScott** [comment]
Thanks, @cartazio!

**Pull request merged**